packages/utils/src/dynamicLB.ts

```diff
@@ -0,0 +1,299 @@
+import {
+  type EIP1193Parameters,
+  TimeoutError,
+  type Transport,
+  type TransportConfig,
+  createTransport,
+} from "viem";
+import {
+  type PromiseWithResolvers,
+  promiseWithResolvers,
+} from "./utils/promiseWithResolvers.js";
+
+export const dynamicLB = (_transports: Transport[]): Transport => {
+  const RPC_USAGE = Array.from({ length: _transports.length }, () => ({
+    s: 0,
+    l: 0,
+  })) as { s: number; l: number }[];
+
+  return ({ chain, retryCount, timeout }) => {
+    const INITIAL_RETRY_DELAY = 2_000;
+    const MAX_RETRY_DELAY = 30_000;
+    const BACKOFF_FACTOR = 2;
+    const DEFAULT_TIMEOUT = 10_000;
+    const LATENCY_WINDOW_SIZE = 500;
+
+    // RPS management
+    const INITIAL_MAX_RPS = 50;
+    const MIN_RPS = 1;
+    const MAX_RPS = 500;
+    const RPS_INCREASE_FACTOR = 1.2;
+    const RPS_DECREASE_FACTOR = 0.7;
+    const SUCCESS_WINDOW_SIZE = 100;
+
+    const queue: [EIP1193Parameters, PromiseWithResolvers<unknown>][] = [];
+
+    const buckets = _transports.map((transport, index) => ({
+      index,
+      retryDelay: 0,
+
+      activeConnections: 0,
+      isActive: true,
+      isJustActivated: false,
+
+      latencies: [] as number[],
+      expectedLatency: 0,
+      requests: [] as number[],
```

Comment:
It could be simpler to define an object that tracks all recent requests in a one object, and then compute expected latency just in time with a function. In a previous effort I tried this, where each element in the array represents a second.

```ts
  requests: {
    timestamp: number;
    response: number;
    count: number;
    errors: number;
  }[];
```

This has the added benefit of letting up figure out the historical requests per second for a bucket.